chainer/training/extensions/variable_unchain.py

```diff
@@ -0,0 +1,32 @@
+from chainer import configuration
+from chainer.training import extension
+from chainer import variable
+import six
+
+
+class unchain_variables(extension.Extension):
+    """Trainer extension to unchain all comptational graphs.
+
+    This extenstion unchain all comptational graphs after all extensions is
+    run to avoid any memory problem.
+    It observes the previous ``chainer.config.keep_graph_on_report`` flag.
+    The extension is enable while the flag is turned on.
+
+    """
+    priority = 0
+
+    def __init__(self):
+        self._prev_flag = None
+
+    def initialize(self, _):
+        self._prev_flag = configuration.config.keep_graph_on_report
+
+    def trigger(self, _):
+        flag = self._prev_flag
+        self._prev_flag = configuration.config.keep_graph_on_report
+        return flag
+
+    def __call__(self, trainer):
+        for var in six.itervalues(trainer.observation):
+            if isinstance(var, variable.Variable):
+                var.unchain_backward()
```

Comment:
A minor point: isn't it enough to call `unchain` instead of `unchain_backward` here, if the motivation is to release the graph kept by the observation?

Reply:
I took the feature as a last resort when an extension has some memory issue that a user cannot debug by themselves.  `unchain_backward` looks suitable here to me just because it's more powerful.

Reply:
As you say, this PR is for some memory issue.
I added this mention to the explanation, thank you.